src/Cli/dotnet/commands/dotnet-new/MSBuildEvaluation/ProjectCapabilityConstraint.cs

```diff
@@ -0,0 +1,153 @@
+                }
+                catch (Exception e)
+                {
+                    throw new Exception("Invalid constraint configuration.", e);
+                }
+
+                IEnumerable<string> configuredCapabilties;
+
+                if (token.Type == JTokenType.String)
+                {
+                    string? configuredCapability = token.Value<string>();
+                    if (string.IsNullOrWhiteSpace(configuredCapability))
+                    {
+                        throw new Exception("Invalid constraint configuration: arguments should not contain empty values.");
+                    }
+                    configuredCapabilties = new[] { configuredCapability };
+                }
+                else
+                {
+                    if (token is not JArray array)
+                    {
+                        throw new Exception("Invalid constraint configuration: arguments should be either a JSON array or a string.");
+                    }
+
+                    configuredCapabilties = array.Values<string>().Select(value =>
+                    {
+                        if (string.IsNullOrWhiteSpace(value))
+                        {
+                            throw new Exception("Invalid constraint configuration: arguments should not contain empty values.");
+                        }
+                        return value;
+                    });
+                }
+
+                if (_evaluationResult.Status == MSBuildEvaluationResult.EvalStatus.NoProjectFound)
+                {
+                    return TemplateConstraintResult.CreateRestricted(this, "No project found.", "This template can only be created inside the project.");
+                }
+                if (_evaluationResult.Status == MSBuildEvaluationResult.EvalStatus.MultipleProjectFound)
+                {
+                    return TemplateConstraintResult.CreateRestricted(this, $"Multiple project(s) found: {string.Join (", ", _evaluationResult.ProjectPath)}", "Specify the project to use.");
+                }
+                if (_evaluationResult.Status == MSBuildEvaluationResult.EvalStatus.NoRestore)
+                {
+                    return TemplateConstraintResult.CreateRestricted(this, $"The project is not restored.", $"Run 'dotnet restore {_evaluationResult.ProjectPath}' to restore the project.");
+                }
+                if (_evaluationResult.Status == MSBuildEvaluationResult.EvalStatus.Failed || _evaluationResult.Status == MSBuildEvaluationResult.EvalStatus.NotEvaluated || _evaluationResult.EvaluatedProject == null)
+                {
+                    return TemplateConstraintResult.CreateRestricted(this, $"Failed to evaluate project context: {_evaluationResult.ErrorMessage}");
+                }
+
+                foreach (string capability in configuredCapabilties)
+                {
+                    if (!_projectCapabilities.Contains(capability, StringComparer.OrdinalIgnoreCase))
```

Comment:
It looks like this can't handle capability expressions like AppliesTo, and is simply an AND operator

https://docs.microsoft.com/en-us/visualstudio/extensibility/appliesto-element-visual-studio-templates?view=vs-2022
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.vsprojectcapabilityexpressionmatcher?view=visualstudiosdk-2022